File: src/nodejs/src/repositories/CachedPatientRepository.ts
import { CacheEntryOptions, defaultCacheOptions } from "../config/CacheEntryOptions";
import { IPatientRepository } from "./IPatientRepository";
import { Patient } from "../entities/Patient";
import logger from "../utils/logger";
import redisClient from "../utils/redisClient";
import { JsonUtils } from "../utils/jsonUtils";

/**
 * Decorator for IPatientRepository that adds Redis caching functionality.
 * Implements the Decorator pattern to wrap an existing repository with caching capabilities.
 */
export class CachedPatientRepository implements IPatientRepository {
  /**
   * Maximum number of retry attempts for cache operations.
   */
  private readonly maxRetries = 3;

  /**
   * Creates a new instance of CachedPatientRepository.
   * @param innerRepository - The repository to decorate with caching.
   * @param cacheOptions - Optional cache configuration options.
   */
  constructor(
    private innerRepository: IPatientRepository,
    private cacheOptions: CacheEntryOptions = defaultCacheOptions
  ) {
    this.innerRepository = innerRepository;
    this.cacheOptions = cacheOptions;
  }

  /**
   * Retrieves a patient by its ID, using cache when available.
   * @param id - The patient ID.
   * @returns The patient if found, null otherwise.
   */
  async getPatientAsync(id: number): Promise<Patient | null> {
    const key = this.getExistenceCacheKey(id);

    try {
      const cachedData = await redisClient.get(key);

      if (cachedData) {
        logger.debug(`Cache hit for patient: ${id}`);
        const data = JsonUtils.deserialize<Patient>(cachedData);
        if (data) {
          return Object.assign(new Patient(), data);
        }
        return null;
      }

      logger.debug(`Cache miss for patient: ${id}`);
    } catch (error) {
      logger.warn(`Failed to read from cache for patient ${id}:`, error);
    }

    const patient = await this.innerRepository.getPatientAsync(id);

    if (patient) {
      await this.saveToCacheAsync(patient);
    }

    return patient;
  }

  /**
   * Checks if a patient exists by ID, using cache when available.
   * @param id - The patient ID.
   * @returns True if the patient exists, false otherwise.
   */
  async isExistPatientAsync(id: number): Promise<boolean> {
    const key = this.getExistenceCacheKey(id);

    try {
      const exists = await redisClient.exists(key);
      if (exists === 1) {
        logger.debug(`Cache hit for patient existence: ${id}`);
        return true;
      }
    } catch (error) {
      logger.warn(`Failed to check cache for patient existence ${id}:`, error);
    }

    const exists = await this.innerRepository.isExistPatientAsync(id);
    return exists;
  }

  /**
   * Retrieves a paginated list of patients within a date range.
   * Note: This method bypasses cache due to complex query nature.
   * @param startTime - The start of the date range.
   * @param endTime - The end of the date range.
   * @param pageNumber - The page number to retrieve.
   * @param pageSize - The number of items per page.
   * @returns A tuple containing the patients array and total count.
   */
  async getPatientsAsync(
    startTime: Date,
    endTime: Date,
    pageNumber: number,
    pageSize: number
  ): Promise<[Patient[], number]> {
    return this.innerRepository.getPatientsAsync(startTime, endTime, pageNumber, pageSize);
  }

  /**
   * Retrieves a patient by name.
   * Note: This method bypasses cache due to non-ID lookup.
   * @param name - The patient name.
   * @returns The patient if found, null otherwise.
   */
  async getPatientByNameAsync(name: string): Promise<Patient | null> {
    return this.innerRepository.getPatientByNameAsync(name);
  }

  /**
   * Adds a new patient to the repository.
   * @param patient - The patient to add.
   * @returns The added patient with generated ID.
   */
  async addAsync(patient: Patient): Promise<Patient> {
    return this.innerRepository.addAsync(patient);
  }

  /**
   * Gets the cache key for a patient's existence.
   * @param id - The patient ID.
   * @returns The cache key.
   */
  private getExistenceCacheKey(id: number): string {
    return `patient:${id}`;
  }

  /**
   * Saves a patient to the cache with retry logic.
   * @param patient - The patient to cache.
   */
  private async saveToCacheAsync(patient: Patient): Promise<void> {
    const key = this.getExistenceCacheKey(patient.id);
    const value = JSON.stringify(patient);
    const ttl = this.getTtlInSeconds();

    for (let attempt = 1; attempt <= this.maxRetries; attempt++) {
      try {
        if (ttl > 0) {
          await redisClient.setex(key, ttl, value);
        } else {
          await redisClient.set(key, value);
        }
        logger.debug(`Cached patient: ${patient.id}`);
        return;
      } catch (error) {
        const delay = Math.pow(2, attempt) * 1000 + Math.random() * 1000;
        logger.warn(
          `Cache write failed for patient ${patient.id}. Attempt ${attempt}/${this.maxRetries}. Retrying in ${delay}ms`,
          error
        );

        if (attempt < this.maxRetries) {
          await this.delayAsync(delay);
        } else {
          logger.error(`Failed to cache patient ${patient.id} after ${this.maxRetries} attempts`, error);
        }
      }
    }
  }

  /**
   * Calculates the TTL (Time To Live) in seconds based on cache options.
   * @returns The TTL in seconds.
   */
  private getTtlInSeconds(): number {
    const { slidingExpirationMinutes, absoluteExpirationRelativeToNowMinutes } = this.cacheOptions;

    // Use the smaller of the two expiration times if both are set
    if (slidingExpirationMinutes && absoluteExpirationRelativeToNowMinutes) {
      return Math.min(slidingExpirationMinutes, absoluteExpirationRelativeToNowMinutes) * 60;
    }

    if (slidingExpirationMinutes) {
      return slidingExpirationMinutes * 60;
    }

    if (absoluteExpirationRelativeToNowMinutes) {
      return absoluteExpirationRelativeToNowMinutes * 60;
    }

    return 0;
  }

  /**
   * Sleep for a specified duration.
   * @param ms - Duration in milliseconds.
   */
  private delayAsync(ms: number): Promise<void> {
    return new Promise((resolve) => setTimeout(resolve, ms));
  }
}

File: src/python/app/repositories/caches/cached_patient_repository.py
"""Cached patient repository with Redis decorator."""

import json
import logging
from app.configs.cache_config import cache_settings
from app.entities import Order
from app.entities import Patient
from app.repositories.patient_repository import PatientRepository
from datetime import datetime
from typing import Optional, Tuple, List
import redis.asyncio as redis

logger = logging.getLogger(__name__)


class CachedPatientRepository:
    """Decorator for PatientRepository that adds Redis caching."""

    def __init__(self, inner_repository: PatientRepository, redis_client: redis.Redis):
        """
        Initialize cached repository.

        Args:
            inner_repository: The inner patient repository to which calls are delegated
            redis_client: Redis client for caching
        """
        self._inner_repository = inner_repository
        self._redis = redis_client
        self._ttl = cache_settings.cache_ttl_seconds

    @staticmethod
    def _get_existence_cache_key(patient_id: int) -> str:
        """Get cache key for a patient."""
        return f"patient:{patient_id}"

    async def _save_to_cache(self, patient: Patient) -> None:
        """Save patient to cache."""
        try:
            key = self._get_existence_cache_key(patient.id)
            patient_dict = {
                "id": patient.id,
                "name": patient.name,
                "created_at": patient.created_at.isoformat(),
                "updated_at": patient.updated_at.isoformat(),
                "orders": [
                    {
                        "id": order.id,
                        "message": order.message,
                        "patient_id": order.patient_id,
                        "created_at": order.created_at.isoformat(),
                        "updated_at": order.updated_at.isoformat(),
                    }
                    for order in patient.orders
                ],
            }
            await self._redis.setex(key, self._ttl, json.dumps(patient_dict))
            logger.debug("Saved patient %s to cache", patient.id)
        except Exception as e:
            logger.error("Error occurred while saving patient to cache with key %s: %s", key, str(e))

    async def get_patient(self, patient_id: int) -> Optional[Patient]:
        """Get patient by ID (with caching)."""
        key = self._get_existence_cache_key(patient_id)

        try:
            cached_data = await self._redis.get(key)
            if cached_data:
                logger.debug("Patient %s retrieved from cache", patient_id)
                patient_dict = json.loads(cached_data)
                patient = Patient(
                    id=patient_dict["id"],
                    name=patient_dict["name"],
                    created_at=datetime.fromisoformat(patient_dict["created_at"]),
                    updated_at=datetime.fromisoformat(patient_dict["updated_at"]),
                )

                patient.orders = [
                    Order(
                        id=order_dict["id"],
                        message=order_dict["message"],
                        patient_id=order_dict["patient_id"],
                        created_at=datetime.fromisoformat(order_dict["created_at"]),
                        updated_at=datetime.fromisoformat(order_dict["updated_at"]),
                    )
                    for order_dict in patient_dict.get("orders", [])
                ]

                return patient
        except Exception as e:
            logger.warning("Failed to get patient from cache: %s", str(e))

        patient = await self._inner_repository.get_patient(patient_id)
        if patient is not None:
            await self._save_to_cache(patient)
        return patient

    async def is_exist_patient(self, patient_id: int) -> bool:
        """Check if patient exists (with caching)."""
        key = self._get_existence_cache_key(patient_id)

        try:
            if await self._redis.exists(key):
                logger.debug("Patient existence check from cache: %s", patient_id)
                return True
        except Exception as e:
            logger.warning("Failed to check patient existence from cache: %s", str(e))

        exists = await self._inner_repository.is_exist_patient(patient_id)
        return exists

    async def get_patients(
        self, start_time: datetime, end_time: datetime, page_number: int, page_size: int
    ) -> Tuple[List[Patient], int]:
        """
        Get patients within date range with pagination.
        Note: List queries are not cached due to complexity.
        """
        return await self._inner_repository.get_patients(start_time, end_time, page_number, page_size)

    async def get_patient_by_name(self, name: str) -> Optional[Patient]:
        """Get patient by name (no caching for this query)."""
        return await self._inner_repository.get_patient_by_name(name)

    async def add(self, patient: Patient) -> Patient:
        """Add new patient (no cache invalidation needed for new records)."""
        return await self._inner_repository.add(patient)
